packages/api/src/controllers/user.js

```diff
@@ -158,7 +159,143 @@ app.post('/verify', validatePost('user-verification'), async (req, res) => {
   }
 })
 
-// TODO: token must grant you one time access to modify the password.
-// Add password reset tokens as a new record in the database with expiration dates and data.
+app.post(
+  '/password/reset',
+  validatePost('password-reset'),
+  async (req, res) => {
+    const { email, password, resetToken } = req.body
+    const userIds = await req.store.query('user', { email: email })
+    if (userIds.length < 1) {
+      res.status(404)
+      return res.json({ errors: ['user not found'] })
+    }
+    const userId = userIds[0]
+
+    let user = await req.store.get(`user/${userId}`)
+    if (!user) {
+      res.status(404)
+      return res.json({ errors: [`user email ${email} not found`] })
+    }
+
+    const tokens = await req.store.query('password-reset-token', {
+      userId: user.id,
+    })
+
+    if (tokens.length < 1) {
+      res.status(404)
+      return res.json({ errors: ['Password reset token not found'] })
+    }
+
+    let dbResetToken
+    for (let i = 0; i < tokens.length; i++) {
+      const token = await req.store.get(
+        `password-reset-token/${tokens[i]}`,
+        false,
+      )
+
+      if (token.resetToken === resetToken) {
+        dbResetToken = token
+      }
+    }
+
+    if (dbResetToken && dbResetToken.expiration > Date.now() / 1000) {
+      // change user password
+      const [hashedPassword, salt] = await hash(password)
+      await req.store.replace({
+        ...user,
+        password: hashedPassword,
+        salt: salt,
+        emailValid: true,
+      })
+
+      user = await req.store.get(`user/${userId}`)
+
+      // delete all reset tokens associated with user
+      for (const t of tokens) {
+        await req.store.delete(`password-reset-token/${t}`)
+      }
+
+      res.status(201)
+      return res.json(user)
+    } else {
+      res.status(403)
+      return res.json({
+        errors: ['incorrect or expired user validation token'],
+      })
+    }
+  },
+)
+
+app.post(
+  '/password/reset-token',
+  validatePost('password-reset-token'),
+  async (req, res) => {
+    const email = req.body.email
+    const userIds = await req.store.query('user', { email: email })
+    if (userIds.length < 1) {
+      res.status(404)
+      return res.json({ errors: ['user not found'] })
+    }
+    const userId = userIds[0]
```

Comment:
Nitpick, but somewhat more terse syntax for this:
```suggestion
    const [userId] = await req.store.query('user', { email: email })
    if (!userId) {
      res.status(404)
      return res.json({ errors: ['user not found'] })
    }
```